IllegalArgumentException in ContentResolverResourceProvider

Google Play for 2021.03.20

2 occurrences, Android 9 and 11

```
java.lang.IllegalArgumentException: 
  at android.database.DatabaseUtils.readExceptionFromParcel (DatabaseUtils.java:165)
  at android.database.DatabaseUtils.readExceptionFromParcel (DatabaseUtils.java:135)
  at android.content.ContentProviderProxy.query (ContentProviderNative.java:418)
  at android.content.ContentResolver.query (ContentResolver.java:804)
  at android.content.ContentResolver.query (ContentResolver.java:753)
  at android.content.ContentResolver.query (ContentResolver.java:711)
  at org.mapsforge.map.android.rendertheme.ContentResolverResourceProvider.queryDir (ContentResolverResourceProvider.java:139)
  at org.mapsforge.map.android.rendertheme.ContentResolverResourceProvider.buildCacheLevel (ContentResolverResourceProvider.java:96)
  at org.mapsforge.map.android.rendertheme.ContentResolverResourceProvider.refreshCache (ContentResolverResourceProvider.java:175)
  at org.mapsforge.map.android.rendertheme.ContentResolverResourceProvider.<init> (ContentResolverResourceProvider.java:89)
  at cgeo.geocaching.maps.mapsforge.v6.RenderThemeHelper.createThemeFor (RenderThemeHelper.java:186)
  at cgeo.geocaching.maps.mapsforge.v6.RenderThemeHelper.reapplyMapTheme (RenderThemeHelper.java:148)
  at cgeo.geocaching.maps.mapsforge.v6.RenderThemeHelper.lambda$selectMapTheme$0 (RenderThemeHelper.java:238)
  at cgeo.geocaching.maps.mapsforge.v6.RenderThemeHelper.lambda$selectMapTheme$0$RenderThemeHelper (Unknown Source)
  at cgeo.geocaching.maps.mapsforge.v6.-$$Lambda$RenderThemeHelper$yePm3t3WbuFSs5trufBVFWocVl4.onClick (Unknown Source:10)
  at com.android.internal.app.AlertController$AlertParams$3.onItemClick (AlertController.java:1169)
  at android.widget.AdapterView.performItemClick (AdapterView.java:318)
  at android.widget.AbsListView.performItemClick (AbsListView.java:1192)
  at android.widget.AbsListView$PerformClick.run (AbsListView.java:3169)
  at android.os.Handler.handleCallback (Handler.java:874)
  at android.os.Handler.dispatchMessage (Handler.java:100)
  at android.os.Looper.loop (Looper.java:198)
  at android.app.ActivityThread.main (ActivityThread.java:6729)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:876)
```

Similar report?

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:4288)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:4328)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2047)
  at android.os.Handler.dispatchMessage (Handler.java:108)
  at android.os.Looper.loop (Looper.java:166)
  at android.app.ActivityThread.main (ActivityThread.java:7529)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:245)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:921)
Caused by: java.lang.IllegalArgumentException: 
  at android.database.DatabaseUtils.readExceptionFromParcel (DatabaseUtils.java:165)
  at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel (DatabaseUtils.java:146)
  at android.content.ContentProviderProxy.openTypedAssetFile (ContentProviderNative.java:698)
  at android.content.ContentResolver.openTypedAssetFileDescriptor (ContentResolver.java:1422)
  at android.content.ContentResolver.openAssetFileDescriptor (ContentResolver.java:1259)
  at android.content.ContentResolver.openInputStream (ContentResolver.java:979)
  at cgeo.geocaching.storage.ContentStorage.openForRead (ContentStorage.java:368)
  at cgeo.geocaching.storage.ContentStorage.openForRead (ContentStorage.java:355)
  at cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider.createMapFileInputStream (MapsforgeMapProvider.java:327)
  at cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider.access$200 (MapsforgeMapProvider.java:48)
  at cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OfflineMultiMapSource.createTileLayer (MapsforgeMapProvider.java:231)
  at cgeo.geocaching.maps.mapsforge.v6.NewMap.switchTileLayer (NewMap.java:619)
  at cgeo.geocaching.maps.mapsforge.v6.NewMap.initializeLayers (NewMap.java:715)
  at cgeo.geocaching.maps.mapsforge.v6.NewMap.onStart (NewMap.java:710)
  at android.app.Instrumentation.callActivityOnStart (Instrumentation.java:1339)
  at android.app.Activity.performStart (Activity.java:7403)
  at android.app.Activity.performRestart (Activity.java:7477)
  at android.app.Activity.performResume (Activity.java:7482)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:4256)
```

--- a/main/src/cgeo/geocaching/maps/mapsforge/v6/RenderThemeHelper.java
+++ b/main/src/cgeo/geocaching/maps/mapsforge/v6/RenderThemeHelper.java
@@ -175,32 +175,35 @@ public class RenderThemeHelper implements XmlRenderThemeMenuCallback, SharedPref
         final String[] themeIdTokens = theme.id.split(ZIP_THEME_SEPARATOR);
         final boolean isZipTheme = themeIdTokens.length == 2;
 
-        final XmlRenderTheme xmlRenderTheme;
         if (theme.fileInfo == null || theme.fileInfo.uri == null || theme.containingFolder == null) {
-            xmlRenderTheme = null;
-        } else if (!isZipTheme) {
-            if (UriUtils.isFileUri(theme.fileInfo.uri)) {
-                xmlRenderTheme = new ExternalRenderTheme(UriUtils.toFile(theme.fileInfo.uri), this);
+            return null;
+        }
+
+        XmlRenderTheme xmlRenderTheme = null;
+        try {
+            if (!isZipTheme) {
+                if (UriUtils.isFileUri(theme.fileInfo.uri)) {
+                    xmlRenderTheme = new ExternalRenderTheme(UriUtils.toFile(theme.fileInfo.uri), this);
+                } else {
+                    xmlRenderTheme = new ContentRenderTheme(getContentResolver(), theme.fileInfo.uri, this);
+                    xmlRenderTheme.setResourceProvider(new ContentResolverResourceProvider(getContentResolver(), ContentStorage.get().getUriForFolder(theme.containingFolder), true));
+                }
             } else {
-                xmlRenderTheme = new ContentRenderTheme(getContentResolver(), theme.fileInfo.uri, this);
-                xmlRenderTheme.setResourceProvider(new ContentResolverResourceProvider(getContentResolver(), ContentStorage.get().getUriForFolder(theme.containingFolder), true));
-            }
-        } else {
-            //always cache the last used ZipResourceProvider. Check if current one can be reused, if not then reload
-            synchronized (cachedZipMutex) {
-                if (cachedZipProvider == null || !themeIdTokens[0].equals(cachedZipProviderFilename)) {
-                    try {
+                //always cache the last used ZipResourceProvider. Check if current one can be reused, if not then reload
+                synchronized (cachedZipMutex) {
+                    if (cachedZipProvider == null || !themeIdTokens[0].equals(cachedZipProviderFilename)) {
                         //reload
                         cachedZipProvider = new ZipXmlThemeResourceProvider(new ZipInputStream(ContentStorage.get().openForRead(theme.fileInfo.uri)), ZIP_RESOURCE_READ_LIMIT);
                         cachedZipProviderFilename = themeIdTokens[0];
-                    } catch (Exception ex) {
-                        Log.w("Problem loading ZIP Theme '" + theme.fileInfo.uri + "'", ex);
-                        cachedZipProvider = null;
-                        cachedZipProviderFilename = null;
                     }
+                    xmlRenderTheme = cachedZipProvider == null ? null : new ZipRenderTheme(themeIdTokens[1], cachedZipProvider, this);
                 }
-                xmlRenderTheme = cachedZipProvider == null ? null : new ZipRenderTheme(themeIdTokens[1], cachedZipProvider, this);
             }
+        } catch (Exception ex) {
+            Log.w("Problem loading Theme [" + theme.id + "]'" + theme.fileInfo.uri + "'", ex);
+            xmlRenderTheme = null;
+            cachedZipProvider = null;
+            cachedZipProviderFilename = null;
         }
         return xmlRenderTheme;
     }

--- a/main/src/cgeo/geocaching/storage/ContentStorage.java
+++ b/main/src/cgeo/geocaching/storage/ContentStorage.java
@@ -326,7 +326,9 @@ public class ContentStorage {
             //values "wa" (for append) and "rwt" (for overwrite) were tested on SDK21, SDK23, SDk29 and SDK30 using "ContentStorageTest"
             //Note that different values behave differently in different SDKs so be careful before changing them
             return context.getContentResolver().openOutputStream(uri, append ? "wa" : "rwt");
-        } catch (IOException | SecurityException se) {
+        } catch (IOException | SecurityException | IllegalArgumentException se) {
+            //SecurityException is thrown for valid Uri which we have no permission to access
+            //IllegalArgumentException is thrown for invalid Uri (e.g. because a folder/file was deleted meanwhile)
             reportProblem(R.string.contentstorage_err_write_failed, se, uri);
         }
         return null;
@@ -366,7 +368,9 @@ public class ContentStorage {
 
         try {
             return this.context.getContentResolver().openInputStream(uri);
-        } catch (IOException | SecurityException se) {
+        } catch (IOException | SecurityException | IllegalArgumentException se) {
+            //SecurityException is thrown for valid Uri which we have no permission to access
+            //IllegalArgumentException is thrown for invalid Uri (e.g. because a folder/file was deleted meanwhile)
             if (!suppressWarningOnFailure) {
                 reportProblem(R.string.contentstorage_err_read_failed, se, uri);
             }

